core/client/src/client.rs

```diff
@@ -1069,8 +1080,17 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 		// if the block is not a direct ancestor of the current best chain,
 		// then some other block is the common ancestor.
 		if route_from_best.common_block().hash != block {
-			// FIXME: #1442 reorganize best block to be the best chain containing
-			// `block`.
+			let new_best = match select_chain {
+				// find the new best block that is a descendent of the finalized
+				// block. we already have the import lock acquired so there's no
+				// need to re-acquire it in `select_chain`.
+				Some(select_chain) => select_chain.best_containing(block, None, None)?
```

Comment:
If this were `finality_target`, this wouldn't work. Neither would it if `best_containing` were implemented for Polkadot in a way that limited to only chains where availability pieces have been seen.